modules/auth/sso/sspi_windows.go

```diff
@@ -0,0 +1,268 @@
+
+	"github.com/quasoft/websspi"
+	gouuid "github.com/satori/go.uuid"
+)
+
+const (
+	tplSignIn base.TplName = "user/auth/signin"
+)
+
+var (
+	// sspiAuth is a global instance of the websspi authentication package,
+	// which is used to avoid acquiring the server credential handle on
+	// every request
+	sspiAuth *websspi.Authenticator
+
+	// Ensure the struct implements the interface.
+	_ SingleSignOn = &SSPI{}
+)
+
+// SSPI implements the SingleSignOn interface and authenticates requests
+// via the built-in SSPI module in Windows for SPNEGO authentication.
+// On successful authentication returns a valid user object.
+// Returns nil if authentication fails.
+type SSPI struct {
+}
+
+// Init creates a new global websspi.Authenticator object
+func (s *SSPI) Init() error {
+	config := websspi.NewConfig()
+	var err error
+	sspiAuth, err = websspi.New(config)
+	return err
+}
+
+// Free releases resources used by the global websspi.Authenticator object
+func (s *SSPI) Free() error {
+	return sspiAuth.Free()
+}
+
+// IsEnabled checks if there is an active SSPI authentication source
+func (s *SSPI) IsEnabled() bool {
+	sources, err := models.ActiveLoginSources(models.LoginSSPI)
+	if err != nil {
+		log.Warn("Could not get login sources: %v\n", err)
+		return false
+	}
+	return len(sources) > 0
+}
+
+// Priority determines the order in which authentication methods are executed.
+// The lower the priority, the sooner the plugin is executed.
+// The SSPI plugin should be executed last as it returns 401 status code
+// if negotiation fails or should continue, which would prevent other
+// authentication methods to execute at all.
+func (s *SSPI) Priority() int {
+	return 50000
+}
+
+// VerifyAuthData uses SSPI (Windows implementation of SPNEGO) to authenticate the request.
+// If authentication is successful, returs the corresponding user object.
+// If negotiation should continue or authentication fails, immediately returns a 401 HTTP
+// response code, as required by the SPNEGO protocol.
+func (s *SSPI) VerifyAuthData(ctx *macaron.Context, sess session.Store) *models.User {
+	if !s.shouldAuthenticate(ctx) {
+		return nil
+	}
+
+	cfg, err := s.getConfig()
+	if err != nil {
+		log.Error("could not get SSPI config: %v", err)
+		return nil
+	}
+
+	userInfo, outToken, err := sspiAuth.Authenticate(ctx.Req.Request, ctx.Resp)
+	if err != nil {
+		log.Warn("Authentication failed with error: %v\n", err)
+		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+
+		// Include the user login page in the 401 response to allow the user
+		// to login with another authentication method if SSPI authentication
+		// fails
+		addFlashErr(ctx, ctx.Tr("auth.sspi_auth_failed"))
+		ctx.Data["EnableOpenIDSignIn"] = setting.Service.EnableOpenIDSignIn
+		ctx.Data["EnableSSPI"] = true
+		ctx.HTML(401, string(tplSignIn))
+		return nil
+	}
+	if outToken != "" {
+		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+	}
+
+	newSep := cfg.SeparatorReplacement
```

Comment:
Maybe add also option to instead of using replacement separator to not have domain in username. IMHO this should also be default so that `domain\username` -> would become `username`

Reply:
Sounds very interesting, but I need to think of a way to distinguish between other account types (eg. local admin user with username `joe`) from domain account (eg. domain `DOMAIN\joe`, which is not necessary the same joe person).

I am currently using `models.GetUserByName(username)` to find the user object for the relevant `DOMAIN\user` account. May be if I remove the domain part, I could introduce a new `GetUserByNameAndType(username, loginType)` to `models` package that would also make sure that the `LoginType` of the account equals `models.LoginSSPI`.

If that sounds right I will proceed with the change.

Reply:
This is also how all other current auth methods work (ex. LDAP). If you create local user `user1` and user  `user1` with LDAP method authenticates it will be same user as local one, so no need for new method. so as if currently local user would be created domain_username and authenticate using windows auth.

Reply:
More toughts on this.

There can be several domains, with trust between them, for example `company1.local` and `company2.local`. The Gitea instance can run with an account from only one of them, but theoretically it could receive user account names for users of both domains.

We definitely don't want two different users from two different domains to use the same Gitea account, nor do we want to prevent any of two users having identical usernames from different domains to use Gitea.

So, may be, the SSPI method should check which domain the server is joined to, and only remove from usernames that domain name, leaving usernames for other domains prefixed with whatever their domain is.

And as that behaviour might not suit all users, an additional setting in SSPI config will be needed:
- A boolean named `Remove own domain name from usernames`

I personally would set that setting to default to `false`, in order not to remove anything from usernames as that would make it more complex to move the Gitea instance to another domain (One would need to either delete all user accounts or use an SQL script to rename them), if needed.

Reply:
Is it possible to check servers domain name?

Reply:
Also with current code even as same user `domain\user1` and `user1@domain.ltd` would be different users in Gitea

Reply:
> Is it possible to check servers domain name?

Yes, it's possible to check which domain the server running Gittea is joined to.

Reply:
> Also with current code even as same user `domain\user1` and `user1@domain.ltd` would be different users in Gitea

Actually while testing SSPI returned only down-level domain names `domain\user1`. The code is also replacing the `@` separator of UPN, just in case someone manages to configure his environment in such a way that SSPI returns UPN names (`user1@domain.ltd`).

Reply:
Imho it would be ok for first iteration to have option to remove domain name (enabled by default to be in par with ldap auth). And later it could be improved with customizable multiple domain handling

Reply:
ok, will add an option to remove domain name (any), enabled by default

Reply:
Should be fixed with commit [Add option in SSPIConfig for removing of domains from logon names](https://github.com/go-gitea/gitea/pull/8463/commits/7528386332105fb2bb0a244d94de6d85d530b4cd)